File: apps/web/src/i18n/ownership.ts
import type { MessageModule } from "./index";

/**
 * 写归属（H01 §4；Go Host 业务所有权迁移 §2.2、§2.11）：只读横幅、保存指示灯，
 * 以及设置页里六个域的归属表。
 *
 * 原因键是服务发来的稳定键（`ownership.switch.pending` 之类），照原样翻译；
 * 没登记的键不显示，免得把内部字符串糊到界面上。
 */
export const ownership: MessageModule = {
  "zh-CN": {
    "ownership.maintenance": "画布正在切换写入方，现在只能查看",
    "ownership.unknown": "尚未确认画布由谁写入，暂不保存",
    "ownership.error": "无法确认画布由谁写入，暂不保存",
    "ownership.recheck": "重新检查",
    "ownership.settings.readonly": "设置正在切换写入方，现在只能查看",
    "ownership.settings.error": "无法确认设置由谁写入，现在只能查看",
    "ownership.settings.unconfirmed": "尚未确认设置由谁写入，这次保存没有发出",
    "ownership.settings.host": "设置的写入已经交给 Host，请重新打开设置页",
    "ownership.save.readonly": "只读，未保存",
    "ownership.title": "写入归属",
    "ownership.note": "切换归属要在本机执行 armadra-host ownership switch。",
    "ownership.loading": "正在读取各域归属…",
    "ownership.unavailable": "读不到归属记录，界面不显示归属状态。",
    "ownership.epoch": "纪元",
    "ownership.domain.canvas": "画布",
    "ownership.domain.settings": "设置",
    "ownership.domain.filesystem": "文件",
    "ownership.domain.session": "会话",
    "ownership.domain.agent": "Agent",
    "ownership.domain.git": "Git",
    "ownership.writer.runtime": "Runtime",
    "ownership.writer.host": "Host",
    "ownership.writer.maintenance": "维护窗口",
    "ownership.writer.unknown": "未知",
    "ownership.writer.error": "读取失败",
    "ownership.reason.ownership.switch.pending": "切换进行中，两侧都拒绝写入",
    "ownership.reason.ownership.switch.unknown":
      "Runtime 的状态没读到，窗口保持打开",
    "ownership.reason.ownership.rollback.exported": "已写出反向导出包，待交回",
  },
  en: {
    "ownership.maintenance":
      "The canvas is switching writers and is view-only right now",
    "ownership.unknown": "Canvas writer not confirmed yet; saving is paused",
    "ownership.error": "Canvas writer could not be confirmed; saving is paused",
    "ownership.recheck": "Check again",
    "ownership.settings.readonly":
      "Settings are switching writers and are view-only right now",
    "ownership.settings.error":
      "The settings writer could not be confirmed; settings are view-only",
    "ownership.settings.unconfirmed":
      "The settings writer is not confirmed; this save was not sent",
    "ownership.settings.host":
      "Settings writes have moved to the Host; reopen this page",
    "ownership.save.readonly": "Read-only, not saved",
    "ownership.title": "Write ownership",
    "ownership.note":
      "Switching is done at this machine with armadra-host ownership switch.",
    "ownership.loading": "Reading each domain's owner…",
    "ownership.unavailable":
      "The ownership record could not be read; no owner is shown.",
    "ownership.epoch": "Epoch",
    "ownership.domain.canvas": "Canvas",
    "ownership.domain.settings": "Settings",
    "ownership.domain.filesystem": "Files",
    "ownership.domain.session": "Sessions",
    "ownership.domain.agent": "Agents",
    "ownership.domain.git": "Git",
    "ownership.writer.runtime": "Runtime",
    "ownership.writer.host": "Host",
    "ownership.writer.maintenance": "Maintenance window",
    "ownership.writer.unknown": "Unknown",
    "ownership.writer.error": "Unreadable",
    "ownership.reason.ownership.switch.pending":
      "A switch is open; both sides refuse writes",
    "ownership.reason.ownership.switch.unknown":
      "The Runtime's state was not read; the window stays open",
    "ownership.reason.ownership.rollback.exported":
      "The reverse export is written; the epoch has not moved back yet",
  },
};

File: apps/host/internal/storage/ownership.go
package storage

import (
	"context"
	"database/sql"
	"errors"
	"slices"
)

// Write ownership of a business domain (host protocol design §4, step 5;
// Go Host 业务所有权迁移 §2.2).
//
// There is one row per domain and six domains. Each record answers a single
// question — which process may write that domain right now, and under which
// epoch — and it answers it explicitly. Nothing here infers ownership from a
// missing row: an absent record is ErrNotFound, never "the Runtime presumably
// still owns it".
//
// The record deliberately lives outside entities/events. It is not business
// content, so it must never be published to a client as a business change and
// must never be replayed out of the event outbox.

const (
	OwnershipDomainCanvas     = "canvas"
	OwnershipDomainSettings   = "settings"
	OwnershipDomainFilesystem = "filesystem"
	OwnershipDomainSession    = "session"
	OwnershipDomainAgent      = "agent"
	OwnershipDomainGit        = "git"

	OwnerRuntime = "runtime"
	OwnerHost    = "host"

	// A settled record is authoritative. The two transitional phases mean a
	// maintenance window is open: both sides refuse writes until the phase is
	// resolved, and a crash mid-switch is therefore visible rather than
	// resolving itself into two processes that both believe they own the data.
	OwnershipSettled     = "settled"
	OwnershipSwitching   = "switching"
	OwnershipRollingBack = "rolling_back"
)

type Ownership struct {
	Domain, Owner, Phase string
	Epoch                uint64
	ImportID, ReasonCode string
	// EventSequence is the Host's event watermark when this record was last
	// settled. A rollback compares it with the current watermark to tell
	// whether the Host published canvas changes a reverse migration would have
	// to carry back, instead of assuming it published none.
	EventSequence            uint64
	Revision                 uint64
	CreatedAtMS, UpdatedAtMS int64
}

// OwnershipDomains lists every domain in switch order, which is also the order
// dependencies are checked in. It is the only place the set is written down.
var OwnershipDomains = []string{
	OwnershipDomainCanvas,
	OwnershipDomainSettings,
	OwnershipDomainFilesystem,
	OwnershipDomainSession,
	OwnershipDomainAgent,
	OwnershipDomainGit,
}

// ValidOwnershipDomain reports whether a name is one of the six. The migration
// constrains the column the same way; refusing here keeps a caller's mistake a
// request error instead of a database one.
func ValidOwnershipDomain(value string) bool {
	return slices.Contains(OwnershipDomains, value)
}

func validOwner(value string) bool { return value == OwnerRuntime || value == OwnerHost }
func validPhase(value string) bool {
	return value == OwnershipSettled || value == OwnershipSwitching || value == OwnershipRollingBack
}

func (o Ownership) validate() error {
	if !ValidOwnershipDomain(o.Domain) || !validOwner(o.Owner) || !validPhase(o.Phase) {
		return ErrInvalid
	}
	if o.Epoch == 0 || !textValid(o.ImportID, 128, true) || !textValid(o.ReasonCode, 64, true) {
		return ErrInvalid
	}
	if o.CreatedAtMS <= 0 || o.UpdatedAtMS <= 0 {
		return ErrInvalid
	}
	if _, err := signed(o.Epoch); err != nil {
		return err
	}
	return nil
}

const ownershipColumns = "SELECT domain,owner,epoch,phase,import_id,reason_code,event_sequence,revision,created_at_ms,updated_at_ms FROM write_ownership WHERE domain=?"

func scanOwnership(row scanner) (Ownership, error) {
	var record Ownership
	var epoch, sequence, revision int64
	err := row.Scan(&record.Domain, &record.Owner, &epoch, &record.Phase, &record.ImportID, &record.ReasonCode, &sequence, &revision, &record.CreatedAtMS, &record.UpdatedAtMS)
	if errors.Is(err, sql.ErrNoRows) {
		return Ownership{}, ErrNotFound
	}
	if err != nil {
		return Ownership{}, err
	}
	if epoch < 1 || revision < 1 || sequence < 0 {
		return Ownership{}, ErrCorrupt
	}
	record.Epoch = uint64(epoch)
	record.EventSequence = uint64(sequence)
	record.Revision = uint64(revision)
	if record.validate() != nil {
		return Ownership{}, ErrCorrupt
	}
	return record, nil
}

// Ownership reads one domain's record. ErrNotFound means no switch has ever
// been recorded on this Host for that domain, which the caller reports as such.
func (s *Store) Ownership(ctx context.Context, domain string) (Ownership, error) {
	if !ValidOwnershipDomain(domain) {
		return Ownership{}, ErrInvalid
	}
	return scanOwnership(s.db.QueryRowContext(ctx, ownershipColumns, domain))
}

// PutOwnership writes the record under revision CAS. Expected revision 0 means
// "never recorded". The epoch is monotonic across every transition: a phase may
// move forward and back, but an epoch never decreases, so a stale controller
// cannot re-apply an older handoff after a newer one landed.
func (s *Store) PutOwnership(ctx context.Context, record Ownership, expectedRevision uint64) (Ownership, error) {
	if err := record.validate(); err != nil {
		return Ownership{}, err
	}
	next, err := signed(expectedRevision + 1)
	if err != nil {
		return Ownership{}, err
	}
	epoch, err := signed(record.Epoch)
	if err != nil {
		return Ownership{}, err
	}
	sequence, err := signed(record.EventSequence)
	if err != nil {
		return Ownership{}, err
	}
	tx, err := s.db.BeginTx(ctx, nil)
	if err != nil {
		return Ownership{}, err
	}
	defer tx.Rollback()
	existing, err := scanOwnership(tx.QueryRowContext(ctx, ownershipColumns, record.Domain))
	switch {
	case errors.Is(err, ErrNotFound):
		if expectedRevision != 0 {
			return Ownership{}, ErrConflict
		}
		record.Revision = 1
		if _, err = tx.ExecContext(ctx, "INSERT INTO write_ownership(domain,owner,epoch,phase,import_id,reason_code,event_sequence,revision,created_at_ms,updated_at_ms) VALUES(?,?,?,?,?,?,?,1,?,?)",
			record.Domain, record.Owner, epoch, record.Phase, record.ImportID, record.ReasonCode, sequence, record.CreatedAtMS, record.UpdatedAtMS); err != nil {
			return Ownership{}, err
		}
		return record, tx.Commit()
	case err != nil:
		return Ownership{}, err
	}
	if existing.Revision != expectedRevision {
		return Ownership{}, ErrConflict
	}
	if record.Epoch < existing.Epoch {
		return Ownership{}, ErrConflict
	}
	record.Revision = uint64(next)
	record.CreatedAtMS = existing.CreatedAtMS
	if err = changed(tx.ExecContext(ctx, "UPDATE write_ownership SET owner=?,epoch=?,phase=?,import_id=?,reason_code=?,event_sequence=?,revision=?,updated_at_ms=? WHERE domain=? AND revision=?",
		record.Owner, epoch, record.Phase, record.ImportID, record.ReasonCode, sequence, next, record.UpdatedAtMS, record.Domain, int64(expectedRevision))); err != nil {
		return Ownership{}, err
	}
	return record, tx.Commit()
}

// AllOwnership reads every recorded domain, keyed by domain name. Domains with
// no record are simply absent: this is the store, and inventing a default here
// would make an unrecorded domain indistinguishable from one this Host really
// did take over. The caller decides what "never recorded" means, once.
func (s *Store) AllOwnership(ctx context.Context) (map[string]Ownership, error) {
	rows, err := s.db.QueryContext(ctx, "SELECT domain,owner,epoch,phase,import_id,reason_code,event_sequence,revision,created_at_ms,updated_at_ms FROM write_ownership")
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	records := make(map[string]Ownership, len(OwnershipDomains))
	for rows.Next() {
		record, err := scanOwnership(rows)
		if err != nil {
			return nil, err
		}
		records[record.Domain] = record
	}
	if err = rows.Err(); err != nil {
		return nil, err
	}
	return records, nil
}
